pkg/controllers/kepler.go

```diff
@@ -168,45 +171,132 @@ func (r KeplerReconciler) updateStatus(ctx context.Context, req ctrl.Request, re
+
+		now := metav1.Now()
+		for i := range kepler.Status.Conditions {
+			kepler.Status.Conditions[i].LastTransitionTime = now
+		}
+
 		return r.Client.Status().Update(ctx, kepler)
 
 	})
 }
 
+func (r KeplerReconciler) updateReconciledStatus(ctx context.Context, k *v1alpha1.Kepler, recErr error) {
+
+	reconciled := v1alpha1.Condition{
+		Type:               v1alpha1.Reconciled,
+		ObservedGeneration: k.Generation,
+		Status:             v1alpha1.ConditionTrue,
+		Reason:             v1alpha1.ReconcileComplete,
+		Message:            "Reconcile succeeded",
+	}
+
+	if recErr != nil {
+		reconciled.Status = v1alpha1.ConditionFalse
+		reconciled.Reason = v1alpha1.ReconcileError
+		reconciled.Message = recErr.Error()
+	}
+
+	k.Status.Conditions = append(k.Status.Conditions, reconciled)
+}
+
+func (r KeplerReconciler) updateAvailableStatus(ctx context.Context, k *v1alpha1.Kepler, recErr error) {
+
+	// get daemonset owned by kepler
+	dset := appsv1.DaemonSet{}
+	key := types.NamespacedName{Name: exporter.DaemonSetName, Namespace: components.Namespace}
+	if err := r.Client.Get(ctx, key, &dset); err != nil {
+		k.Status.Conditions = append(k.Status.Conditions, v1alpha1.Condition{
+			Type:    v1alpha1.Available,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  v1alpha1.DaemonSetNotFound,
+			Message: err.Error(),
+		})
+		return
+	}
+
+	ds := dset.Status
+	k.Status.NumberMisscheduled = ds.NumberMisscheduled
+	k.Status.CurrentNumberScheduled = ds.CurrentNumberScheduled
+	k.Status.DesiredNumberScheduled = ds.DesiredNumberScheduled
+	k.Status.NumberReady = ds.NumberReady
+	k.Status.UpdatedNumberScheduled = ds.UpdatedNumberScheduled
+	k.Status.NumberAvailable = ds.NumberAvailable
+	k.Status.NumberUnavailable = ds.NumberUnavailable
+
+	c := availableCondition(ds)
+	if recErr == nil {
+		c.ObservedGeneration = k.Generation
+	}
+	k.Status.Conditions = append(k.Status.Conditions, c)
+}
+
+func availableCondition(ds appsv1.DaemonSetStatus) v1alpha1.Condition {
+	c := v1alpha1.Condition{Type: v1alpha1.Available}
+
+	// NumberUnavailable: The number of nodes that should be running the daemon pod
+	// and have none of the daemon pod running and available (ready for at least
+	// spec.minReadySeconds)
+	if x := ds.NumberUnavailable; x > 0 {
+		c.Status = v1alpha1.ConditionFalse
+		c.Reason = v1alpha1.DaemonSetUnavailable
+		c.Message = fmt.Sprintf("Kepler pod is not running or available on %d nodes", x)
+		return c
+	}
+
+	// NumberAvailable: The number of nodes that should be running the daemon pod
+	// and have one or more of the daemon pod running and available (ready for at
+	// least spec.minReadySeconds)
+	if ds.NumberAvailable < ds.DesiredNumberScheduled {
+		c.Status = v1alpha1.ConditionUnknown
+		c.Reason = v1alpha1.DaemonSetPartiallyAvailable
+		c.Message = fmt.Sprintf("Only %d of total %d pods are running and available; pending: %d",
+			ds.NumberAvailable, ds.DesiredNumberScheduled, ds.NumberUnavailable)
+		return c
+	}
```

Comment:
will this block ever hit?

Reply:
Good question, I thought this should be a state that happens .. Searching for this got me 
https://github.com/kubernetes/kubectl/blob/master/pkg/polymorphichelpers/rollout_status.go#L108C5-L108C5